cluster/query_executor.go

```diff
@@ -481,42 +482,29 @@ func (e *QueryExecutor) executeSelectStatement(stmt *influxql.SelectStatement, c
 	}
 
 	// Emit rows to the results channel.
-	var writeN int64
-	var emitted bool
-	for {
-		row := em.Emit()
-		if row == nil {
-			// Check if the query was interrupted while emitting.
-			select {
-			case <-closing:
-				return influxql.ErrQueryInterrupted
-			default:
-			}
-			break
-		}
-
-		result := &influxql.Result{
-			StatementID: statementID,
-			Series:      []*models.Row{row},
-		}
+	var emstats emitterStats
+	emitterCh := make(chan emitterStats, 1)
+	done := make(chan struct{})
+	go e.monitorEmitter(em, stmt, statementID, emitterCh, done, results, closing)
+	defer func() { <-done }()
```

Comment:
Yea, the query manager needs to be modified to allow interrupts to send errors. It's an iterative process to include it. You don't necessarily need to attach iterators to QueryTaskInfo though since that's supposed to be a public data only struct. I was going to add a `MonitorQuery` function that would allow running a custom function at certain intervals.

On the query manager itself, I thought about adding it directly to QueryExecutor, but I wanted to separate the task management process of the query from the executor. I also wanted to be able to reuse the query management so it would be easier to test in isolation. In the code right now, the executor has no code dedicated to managing the query. The executor adds the query to the query manager and then never touches it again outside of looking at a channel that determines when the query is closed. The executor exists for the entire lifetime of the query since it's the one handling the execution of the query.